rust/headless-client/src/dns_control/windows.rs

```diff
@@ -104,31 +110,64 @@ fn activate(dns_config: &[IpAddr]) -> Result<()> {
     // using QUIC, HTTP/2, or even HTTP/1.1, and so they won't resolve the DNS
     // again unless you let that connection time out:
     // <https://github.com/firezone/firezone/issues/3113#issuecomment-1882096111>
+    tracing::info!("Activating DNS control...");
+
+    let hklm = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+
+    // If this key exists, our local NRPT rules are ignored and we have to stick
+    // them in with group policies for some reason.
+    let group_policy_key_exists = hklm.open_subkey(group_nrpt_path()).is_ok();
+    tracing::debug!(?group_policy_key_exists);
 
-    tracing::info!("Activating DNS control");
     let dns_config_string = dns_config
         .iter()
         .map(|ip| ip.to_string())
         .collect::<Vec<_>>()
         .join(";");
 
-    let hkcu = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
-    let base = Path::new("SYSTEM")
-        .join("CurrentControlSet")
-        .join("Services")
-        .join("Dnscache")
-        .join("Parameters")
-        .join("DnsPolicyConfig")
-        .join(NRPT_REG_KEY);
+    // It's safe to always set the local rule.
+    let (key, _) = hklm.create_subkey(local_nrpt_path().join(NRPT_REG_KEY))?;
+    set_nrpt_rule(&key, &dns_config_string)?;
+
+    if group_policy_key_exists {
+        let (key, _) = hklm.create_subkey(group_nrpt_path().join(NRPT_REG_KEY))?;
```

Comment:
Probably a TOCTOU gap here, if someone is taking the computer off of Group Policy right as we're activating DNS. Worth fixing before merging maybe, but not before testing

Reply:
Probably such a low chance it might not be worth the added complexity. I think always having the local key set should mean that it will fall back to that if GPO is disabled?

Reply:
Yeah I'll leave it alone for now then and just leave a TODO